datafusion/physical-expr/src/aggregate/count.rs

```diff
@@ -198,16 +198,18 @@ fn null_count_for_multiple_cols(values: &[ArrayRef]) -> usize {
     if values.len() > 1 {
         let result_bool_buf: Option<BooleanBuffer> = values
             .iter()
-            .map(|a| a.nulls())
+            .map(|a| a.logical_nulls())
```

Comment:
I am slightly worried about the need to allocate a new null buffer each time, even for arrays when we could just use the exising one

This is particularly concerning given this is on the critical path in aggregates

I reviewed the `logical_nulls` method -- 
https://docs.rs/arrow/latest/arrow/array/trait.Array.html#method.logical_nulls and I see the issue is that it returns an owned Option

What would you think about implemeting a method in `DataFusion` that avoids the copy if it is not necessary, like
```rust
fn logical_nulls(arr: &ArrayRef) -> Cow<'_, Option<BooleanBuffer>> {
  
}
```

That only creates the nulll buffer for `NullArrays`? 

Then we can propose upstreaming that back to arrow-rs to avoid the potential performance issue

I know the Cow thing is not always the easiest to make happen -- if you need help I can try and find time to help code it up.

Reply:
I'm not sure how I would implement this outside of the `Array` trait while ensuring that all cases are covered. Originally I had some branching logic based on the datatype but removed it after the discussion here: https://github.com/apache/arrow-datafusion/pull/8511#discussion_r1423733822

Reply:
Yeah @alamb in the end `NullBuffer` has `Arc<Bytes>` so it mostly clones this + a few `usize`s etc. While not ideal I don't think it will be very expensive?
https://arrow.apache.org/rust/arrow_buffer/buffer/immutable/struct.Buffer.html

But I like the suggestion of returning a reference or `Cow` in arrow-rs.

Reply:
I am not opposed to this PR, but I would prefer to have the `Cow` thing. Let me see if I can whip it up quickly

Reply:
Here is my proposal for improvement: https://github.com/coralogix/arrow-datafusion/pull/221